tests/0101-fetch-from-follower.cpp

```diff
@@ -0,0 +1,320 @@
+          char* cnt_str_start_ptr = strstr((char *)msg->payload(), "msg=") + 4;
+          test_assert(cnt_str_start_ptr, "expecting 'msg=' in message payload");
+          char* cnt_str_end_ptr = strstr(cnt_str_start_ptr, "\n");
+          test_assert(cnt_str_start_ptr, "expecting '\n' following 'msg=' in message payload");
+          *cnt_str_end_ptr = '\0';
+          int msg_cnt = atoi(cnt_str_start_ptr);
+          test_assert(msg_cnt == cnt, "message consumed out of order");
+          cnt++;
+        }
+        break;
+
+      default:
+        Test::Fail("Consume error: " + msg->errstr());
+        break;
+      }
+
+    delete msg;
+  }
+
+  /* rely on the test timeout to prevent an infinite loop in
+   * the (unlikely) event that the statistics callback isn't
+   * called. */
+  Test::Say("Capturing rxbytes statistics\n");
+  TestEvent2Cb::should_capture_stats = true;
+  while (!TestEvent2Cb::has_captured_stats) {
+    RdKafka::Message *msg = c->consume(tmout_multip(500));
+    delete msg;
+  }
+
+  for (i=0; i<replica_ids.size(); ++i)
+    Test::Say(tostr() << _C_YEL << "rxbytes for replica on broker " << replica_ids[i] << ": " << TestEvent2Cb::rxbytes[replica_ids[i]]
+                      << (replica_ids[i] == leader_id ? " (leader)" : "")
+                      << (replica_ids[i] == client_rack_id ? " (preferred replica)" : "")
+                      << "\n");
+
+  for (i=0; i<replica_ids.size(); ++i)
+    if (replica_ids[i] != client_rack_id)
+      test_assert(TestEvent2Cb::rxbytes[replica_ids[i]] < TestEvent2Cb::rxbytes[client_rack_id],
+                  "rxbytes was not highest on broker corresponding to client.rack.");
+
+  test_assert(TestEvent2Cb::rxbytes[client_rack_id] > msgcnt * msgsize,
+              tostr() << "expecting rxbytes of client.rack broker to be at least " << msgcnt * msgsize
+                      << " but it was " << TestEvent2Cb::rxbytes[client_rack_id]);
+
+  Test::Say("Done\n");
+
+  // Manual test 1:
+  //  - change the lease period from 5 minutes to 5 seconds (modify rdkafka_partition.c)
+  //  - change the max lease grant period from 1 minute to 10 seconds (modify rdkafka_broker.c)
+  //  - add infinite consume loop to the end of this test.
+  //  - observe:
+  //     - the partition gets delegated to the preferred replica.
+  //     - the messages get consumed.
+  //     - the lease expires.
+  //     - the partition is reverted to the leader.
+  //     - the toppar is backed off, and debug message noting the faster than expected delegation to a replica.
+
+  // Manual test 2:
+  //  - same modifications as above.
+  //  - add Test::conf_set(conf, "topic.metadata.refresh.interval.ms", "3000");
+  //  - observe:
+  //     - that metadata being periodically received and not interfering with anything.
+
+  c->close();
+  delete c;
+}
+#endif
+
+extern "C" {
+  int main_0101_fetch_from_follower (int argc, char **argv) {
+    if (!test_getenv("TRIVUP_ROOT", NULL)) {
```

Comment:
This is a bit too broad-edged (and TRIVUP_ROOT can be set outside a trivup environment to point out the path to where it should store its local state (which is bad reuse of an env, I know)).

Since you're already collecting the broker racks in the test itself I suggest you Skip the test from there instead if there are less than 2 racks configured.

Reply:
i'm not actually doing that - need to do a describe configs to get that information.